

####################    samtools    ########################
samtools view -@ 20 -s 0.05 -o sub.bam data.sort.bam
samtools index -@ 20 sub.bam
samtools depth frag1.bam |  awk '{sum+=$3} END { print "frag\t"sum/NR}' >> ../../$i.frag_depth.txt
		



############################################################
# while circulation
while read a  c d b
do
    echo "File:${d}"
done < tmp.txt

####################    bcftools & bedtools    ########################
bcftools view -O z --type snps {input} > {output}"
bcftools view -O z --type indels {input} > {output}"

bcftools view -O v -r $chr phasetransfer.vcf.gz > truth_chroms_${chr}.vcf
bcftools filter -i 'CALL=TP' -Ov -o fDGQX3.vcf calls.vcf.gz

bcftools view unphasing.vcf -Oz -o unphasing.vcf.gz
bcftools index unphasing.vcf.gz
tabix -p vcf $i.gz

# get unphasing.vcf
bcftools isec -n~10 -c none data_gatk_pass_vars.vcf.gz data_hapcut.phased.vcf.gz -w1 > unphasing.vcf

# fp.vcf.gz  and  unphasing or phasing vcf.gz

bcftools isec -n~11 -c none fp.vcf.gz unphasing.vcf.gz -w1 > fp_unphasing.c.vcf
bcftools isec -n~11 -c none fp.vcf.gz data_hapcut.phased.vcf.gz -w1 > fp_phasing.c.vcf

bedtools shuffle -i fp.vcf.bed -g /home/zhenyuluo/work_path/reference/human/hg38.chrom.sizes > random.bed

bedtools intersect -a fp.snp.random.sort.bed -b hg38.trf.sort.bed > fp_R_hg38.trf.bed


############################################################

###############################  gzip  ###############################

压缩保留源文件的方法：
gzip –c filename > filename.gz
Linux解压缩保留源文件的方法：
gunzip –c filename.gz > filename




############################################################


shuf -n 20947 xx.txt > xx.txt
sort -k1,1 -k2,2n hg38.trf.bed > hg38.trf.sort.bed
samtools faidx chr19.GCA_000001405.15_GRCh38_no_alt_analysis_set.fa
bwa index chr19.GCA_000001405.15_GRCh38_no_alt_analysis_set.fa
awk '{print $0,FILENAME}' OFS='\t' $i|sed 's/data_hapblock_//g; s/.all//g' >> 1_all_chr_Block.txt







seqtk sample -s100 ${i}_1.fq.gz 10000 | gzip >${i}_1.fq.gz

seqtk sample -s100 /research/rv-02/Projects/Project_stLFR//V350096781/V350096781_L01_unfiltered.fq.gz 40000 | gzip > V350096781_L01_read_2.fq.gz

makeblastdb -in Sitalica.con -dbtype nucl -out Sitalicagenome

blastn -task blastn -query data.uniq.unmaped.fasta \
-db /home/zhenyuluo/work_path/reference/nt_nr/nt.fa -outfmt 7 \
-evalue 0.00001  -max_target_seqs 1 -num_threads 100 -out data.uniq.unmaped.txt
blastp -query cold_AT_gffclean.fa -db ../Nitab-v4.5_db -outfmt 7 -evalue 0.00001  -max_target_seqs 1 -num_threads 20 -out cold_AT_gffclean.txt

fastp -i data.uniq.unmaped.fastq.gz -o data.uniq.unmaped.fastq.filter.gz \
        -Q --thread=80 --length_required=50
































































































































































































































































































































































































































































































































































































































































































































































































































