{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 本教程是复现[《Different fibroblast subtypes propel spatially defined ileal inflammation through TNFR1 signalling in murine ileitis》](https://www.nature.com/articles/s41467-025-57570-7)这篇文章中的数据分析过程。\n",
    "- 脚本下载：[github](https://github.com/ChrisTzaferis/Iliopoulou_et_Al_Wt_Dare12wk_sc_analysis.git)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step2_InitialSeuratObjects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```r\n",
    "library(Seurat)\n",
    "library(patchwork)\n",
    "library(dplyr)\n",
    "\n",
    "#object1 initialization\n",
    "obj1_counts <- Read10X(\"../GEO_files/PROCESSED/WT12/\")\n",
    "obj1 <- CreateSeuratObject(counts = obj1_counts, project = \"WT12wk\",\n",
    "                           min.cells = 0, min.features = 0)\n",
    "# 返回一个稀疏计数矩阵，其中行代表基因，列代表细胞条形码（barcode），\n",
    "# 值代表该基因在该细胞中检测到的分子数（UMI计数）\n",
    "obj1[[\"percent.mt\"]] <- PercentageFeatureSet(obj1, pattern = \"^mt-\")\n",
    "# 线粒体基因编码的RNA占比。\n",
    "# 线粒体基因高表达通常指示细胞凋亡或损伤。\n",
    "# 小鼠线粒体基因以\"mt-\"开头，人类以\"MT-\"开头\n",
    "obj1$sample <- \"wt-12w\"\n",
    "obj1$genotype <- \"wt\"\n",
    "head(obj1@meta.data)\n",
    "\n",
    "#doublet removal\n",
    "#load scrublet results\n",
    "# 双细胞问题：​​ 微流体实验中两个细胞可能被包裹在同一个液滴中，形成\"双细胞\"，会误导下游分析\n",
    "obj1_doublets <- read.csv(\"../Step1_DoubletRemovalWithScrublet/WT12_scrublet_results.csv\")\n",
    "colnames(obj1_doublets)[1] <- \"Cell_id\"\n",
    "obj1_doublets <- obj1_doublets[, c(\"Cell_id\", \"predicted_doublets\")]\n",
    "\n",
    "metadata_temp1 <- as.data.frame(obj1@meta.data)\n",
    "metadata_temp1$Cell_id <- rownames(metadata_temp1)\n",
    "metadata_temp1 <- left_join(metadata_temp1, obj1_doublets)\n",
    "rownames(metadata_temp1) <- metadata_temp1$Cell_id\n",
    "\n",
    "obj1@meta.data <- metadata_temp1\n",
    "Idents(obj1) <- obj1$predicted_doublets\n",
    "\n",
    "#View QC plots before doublet removal\n",
    "VlnPlot(obj1, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"),\n",
    "        ncol = 3, group.by = \"orig.ident\", pt.size = 0)\n",
    "\n",
    "# 三个关键质控指标：​​\n",
    "# ​nFeature_RNA​：每个细胞检测到的基因数\n",
    "# ​nCount_RNA​：每个细胞的总UMI计数（转录本数）\n",
    "# ​percent.mt​：线粒体基因占比\n",
    "# 质控阈值判断标准\n",
    "# \n",
    "# ​nFeature_RNA过低​：可能为空液滴或死细胞\n",
    "# ​nFeature_RNA过高​：可能为多个细胞形成的双细胞\n",
    "# ​percent.mt过高​：通常>5-10%表示细胞状态不佳\n",
    "\n",
    "#Remove doublets\n",
    "obj1_no_doublets <- subset(obj1, idents = \"False\")\n",
    "\n",
    "obj1 #43119 cells\n",
    "obj1_no_doublets #35887 cells\n",
    "\n",
    "#View QC plots after doublet removal\n",
    "VlnPlot(obj1_no_doublets, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3, group.by = \"orig.ident\", pt.size = 0)\n",
    "saveRDS(obj1_no_doublets, \"../output/wt12_no_doubletsDefaultThreshold_not_filtered.RDS\")\n",
    "\n",
    "#object2 initialization\n",
    "obj2_counts <- Read10X(\"../GEO_files/PROCESSED/DARE12/\")\n",
    "obj2 <- CreateSeuratObject(counts = obj2_counts, project = \"DARE12wk\", \n",
    "                           min.cells = 0, min.features = 0)\n",
    "obj2[[\"percent.mt\"]] <- PercentageFeatureSet(obj2, pattern = \"^mt-\")\n",
    "obj2$sample <- \"dare-12w\"\n",
    "obj2$genotype <- \"dare\"\n",
    "\n",
    "#doublet removal\n",
    "#load scrublet results\n",
    "obj2_doublets <- read.csv(\"../Step1_DoubletRemovalWithScrublet/DARE12_scrublet_results.csv\")\n",
    "colnames(obj2_doublets)[1] <- \"Cell_id\"\n",
    "obj2_doublets <- obj2_doublets[, c(\"Cell_id\", \"predicted_doublets\")]\n",
    "\n",
    "metadata_temp2 <- as.data.frame(obj2@meta.data)\n",
    "metadata_temp2$Cell_id <- rownames(metadata_temp2)\n",
    "metadata_temp2 <- left_join(metadata_temp2, obj2_doublets)\n",
    "rownames(metadata_temp2) <- metadata_temp2$Cell_id\n",
    "\n",
    "obj2@meta.data <- metadata_temp2\n",
    "Idents(obj2) <- obj2$predicted_doublets\n",
    "\n",
    "#View QC plots before doublet removal\n",
    "VlnPlot(obj2, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3, group.by = \"orig.ident\", pt.size = 0)\n",
    "\n",
    "#Remove doublets\n",
    "obj2_no_doublets <- subset(obj2, idents = \"False\")\n",
    "\n",
    "obj2 #25771 cells\n",
    "obj2_no_doublets #22708 cells\n",
    "\n",
    "#View QC plots after doublet removal\n",
    "VlnPlot(obj2_no_doublets, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3, group.by = \"orig.ident\", pt.size = 0)\n",
    "saveRDS(obj2_no_doublets, \"../output/dare12_no_doubletsDefaultThreshold_not_filtered.RDS\")\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 解读Seurat对象 \n",
    "    要把Seurat对象，当做一个数据库，包含多种数据\n",
    "    其中，最重要的就是包含了多个表达矩阵和细胞注释信息（类似于临床信息）\n",
    "    当需要用到表达矩阵或细胞注释信息时，会使用默认信息\n",
    "    不想要使用默认信息，就必须指定或修改默认信息\n",
    "    用@,$符号依次取下层，也可以[[]]\n",
    "    meta.data：    储存着细胞注释信息（类似于临床信息）\n",
    "    active.ident： 储存着默认的细胞注释信息（类似于临床信息）\n",
    "    active.assay： 储存着默认的矩阵名\n",
    "    assays：       储存着表达矩阵\n",
    "    counts：       存储原始数据，是稀疏矩阵\n",
    "    data：         存储Normalize() 规范化的data\n",
    "    scale.data：   存储 ScaleData()缩放后的data\n",
    "    SCT：          储存SCT标准化之后的data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 其他格式的读取"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### H5格式"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```r\n",
    "setwd(\"GSE199866\")\n",
    "fs=list.files(pattern = '.h5')\n",
    "fs\n",
    "sceList = lapply(fs, function(x){\n",
    "    a=Read10X_h5(x)\n",
    "    p=strsplit(x,'',simplify = T)[,1]\n",
    "    sce <- CreateSeuratObject(a,project = p ,min.features = 200,min.cells = 3)\n",
    "})\n",
    "folders = substr(fs,1,10)\n",
    "folders\n",
    "sce.big <- merge(sceList[[1]],  #使用merge函数进行合并\n",
    "    y = sceList[-1],\n",
    "    add.cell.ids = folders)\n",
    "pbmc = JoinLayers(sce.big)#将Layers融合\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R数据文件(RDS/RDATA文件)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```r\n",
    "setwd(\"F:/4scRNA/1code/3/3\")\n",
    "load(file =\"1pbmc.RData\")#读取RDATA文件\n",
    "pbmc2 = readRDS(\"1pbmc.rds\")#读取RDS文件"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step3_SeuratWorkflowPerSample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```r\n",
    "rm(list=ls())\n",
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "# remotes::install_github(\"samuel-marsh/scCustomize@v1.1.3\")\n",
    "library(scCustomize)\n",
    "\n",
    "#load object\n",
    "obj <- readRDS(\"../Step2_InitialSeuratObjects/wt12_no_doubletsDefaultThreshold_not_filtered.RDS\")\n",
    "obj\n",
    "\n",
    "#QC filtering\n",
    "VlnPlot(obj, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"))\n",
    "# p <- VlnPlot(obj, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"))\n",
    "# # 保存为PNG（位图，适合网页展示）\n",
    "# ggsave(plot = p, filename = \"QC_VlnPlot.png\", width = 10, height = 6, dpi = 300)\n",
    "\n",
    "obj_filtered <- subset(obj, subset = nFeature_RNA > 800 & \n",
    "                         nFeature_RNA < 7500 & percent.mt < 5)\n",
    "# nFeature_RNA > 800​：确保细胞有足够的基因表达信息\n",
    "# ​nFeature_RNA < 7500​：排除可能的多细胞或双细胞\n",
    "# ​percent.mt < 5​：排除线粒体基因污染严重的低质量细胞\n",
    "VlnPlot(obj_filtered, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"))\n",
    "obj_filtered\n",
    "\n",
    "#wt: \n",
    "# before filtering: 35428 cells\n",
    "# nFeature_RNA > 800 & nFeature_RNA < 7500 & percent.mt < 5: 24249 cells\n",
    "\n",
    "#normalization\n",
    "obj_filtered <- NormalizeData(obj_filtered)\n",
    "# 作用​：消除细胞间测序深度的技术差异。\n",
    "# ​原理​：默认使用\"LogNormalize\"方法，将每个细胞的基因表达量除以该细胞的总UMI数，乘以缩放因子10,000，然后进行对数转换\n",
    "# 。这使得不同细胞间的表达量具有可比性。\n",
    "\n",
    "#HVGs\n",
    "obj_filtered <- FindVariableFeatures(obj_filtered, selection.method = \"mvp\")\n",
    "# 作用​：识别在细胞间表达变异最大的基因，用于下游分析。\n",
    "# ​原理​：高变基因通常包含区分不同细胞类型的生物信号，\n",
    "# 聚焦这些基因可以提高分析效率和准确性。\n",
    "# \"mvp\"方法基于平均表达量和离散度来选择基因\n",
    "\n",
    "# Identify the 10 most highly variable genes\n",
    "top10 <- head(VariableFeatures(obj_filtered), 10)\n",
    "\n",
    "# plot variable features with and without labels\n",
    "plot1 <- VariableFeaturePlot(obj_filtered)\n",
    "plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)\n",
    "plot1 + plot2\n",
    "\n",
    "#scaling\n",
    "#all.genes <- rownames(obj_filtered)\n",
    "obj_filtered <- ScaleData(obj_filtered, vars.to.regress = c(\"nCount_RNA\", \"percent.mt\"))\n",
    "# 作用​：标准化基因表达量，消除技术变异的影响。\n",
    "# ​原理​：\n",
    "# 将每个基因的表达量转换为z-score（均值为0，方差为1）\n",
    "# vars.to.regress参数回归去除nCount_RNA和percent.mt的影响，\n",
    "# 减少测序深度和线粒体污染带来的技术偏差\n",
    "# 为PCA分析做准备，确保高表达基因不会主导分析结果\n",
    "\n",
    "#PCA\n",
    "obj_filtered <- RunPCA(obj_filtered, features = VariableFeatures(object = obj_filtered))\n",
    "ElbowPlot(obj_filtered, ndims = 50)\n",
    "\n",
    "#Clustering\n",
    "obj_filtered <- FindNeighbors(obj_filtered, dims = 1:20)\n",
    "obj_filtered <- FindClusters(obj_filtered, resolution = 0.01)\n",
    "# FindNeighbors：基于PCA空间构建K近邻图，连接相似的细胞\n",
    "# FindClusters：应用Louvain算法识别细胞群落，resolution参数控制聚类粒度（值越小聚类越少）\n",
    "\n",
    "#UMAP\n",
    "obj_filtered <- RunUMAP(obj_filtered, dims = 1:20)\n",
    "DimPlot(obj_filtered, reduction = \"umap\", label = T)\n",
    "# 作用​：非线性降维可视化，保留细胞间的全局和局部结构关系。\n",
    "# ​原理​：UMAP比t-SNE能更好地保留数据的全局结构，用于直观展示细胞亚群的分布\n",
    "\n",
    "#annotation of WT\n",
    "new.cluster.ids <- c(\"Lymphoid\", \"Stroma\", \"Lymphoid\", \"Myeloid\", \"Stroma\", \"Stroma\")\n",
    "names(new.cluster.ids) <- levels(obj_filtered)\n",
    "obj_filtered <- RenameIdents(obj_filtered, new.cluster.ids)\n",
    "# 作用​：基于已知的标记基因将聚类结果转换为有生物学意义的细胞类型。\n",
    "# ​注释依据​：\n",
    "# ​Lymphoid​：免疫细胞谱系（如T细胞、B细胞）\n",
    "# ​Myeloid​：髓系免疫细胞（如巨噬细胞、树突状细胞）\n",
    "# ​Stroma​：基质细胞（如成纤维细胞\n",
    "#save\n",
    "saveRDS(obj_filtered, \"WT12_weeks_before_integration_nF_800.RDS\") \n",
    "\n",
    "#used in Fig1b\n",
    "FeaturePlot(obj_filtered, features = \"Itgax\", order = T, label = T, cols = c(\"grey70\", \"red\"))\n",
    "FeaturePlot(obj_filtered, features = \"Itgam\", order = T, label = T, cols = c(\"grey70\", \"red\"))\n",
    "FeaturePlot(obj_filtered, features = \"Ptprc\", order = T, label = T, cols = c(\"grey70\", \"red\"))\n",
    "FeaturePlot(obj_filtered, features = \"Col1a1\", order = T, label = T, cols = c(\"grey70\", \"red\"))\n",
    "# 验证\n",
    "# Itgax​（CD11c）：树突状细胞标记\n",
    "# ​Itgam​（CD11b）：髓系细胞标记\n",
    "# ​Ptprc​（CD45）：白细胞共同标记\n",
    "# ​Col1a1​：胶原蛋白，基质细胞标记\n",
    "\n",
    "DimPlot_scCustom(obj_filtered, colors_use = c(\"purple\", \"coral\", \"grey\"), pt.size = 0.8, label.size = 7) + \n",
    "  theme(axis.line = element_blank(), \n",
    "        axis.text = element_blank(), \n",
    "        axis.ticks = element_blank(),\n",
    "        axis.title = element_blank()\n",
    "        )\n",
    "#  分析流程总结\n",
    "# 这个流程体现了单细胞分析的标准范式：\n",
    "# 质控过滤→标准化→特征选择→降维聚类→细胞注释\n",
    "# 每个步骤都为后续分析奠定基础，确保结果的生物学可靠性\n",
    "# 这种分析能够揭示细胞异质性，识别稀有细胞类型，\n",
    "# 为理解组织微环境的细胞组成和功能状态提供重要见解。\n",
    "\n",
    "# Q:高变基因识别之前不对数据进行缩放和批次效应矫正\n",
    "# 数据标准化在高变基因识别之前；\n",
    "#     消除细胞间测序深度的技术差异，使表达量可比。为准确评估基因变异提供公平的“起跑线”。\n",
    "# 而数据缩放和批次效应矫正在后：避免对高变基因选择引入偏差。\n",
    "#     将基因表达值转换为Z-score，为线性降维（如PCA）做准备。\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step4_Analysis_of_Stroma_Lymphoid_Myeloid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```r\n",
    "rm(list=ls())\n",
    "library(Seurat)\n",
    "\n",
    "wt12 <- readRDS(\"../Step3_SeuratWorkflowPerSample/WT12_weeks_before_integration_nF_800.RDS\")\n",
    "dare12 <- readRDS(\"../Step3_SeuratWorkflowPerSample/DARE12_weeks_before_integration_nF_800.RDS\")\n",
    "\n",
    "analyzeAndSaveObjects <- function(seu_obj, supercluster, nPCs, resolution, sampleName)\n",
    "{\n",
    "  obj_filtered <- subset(seu_obj, idents = supercluster)\n",
    "  # 从完整的Seurat对象中，根据细胞身份（idents）提取出指定大类（如 \"Myeloid\"）的所有细胞，\n",
    "  # 创建一个新的、只包含该大类细胞的Seurat对象\n",
    "  \n",
    "  #normalization\n",
    "  obj_filtered <- NormalizeData(obj_filtered)\n",
    "  \n",
    "  #HVGs\n",
    "  obj_filtered <- FindVariableFeatures(obj_filtered, selection.method = \"mvp\")\n",
    "  \n",
    "  # Identify the 10 most highly variable genes\n",
    "  top10 <- head(VariableFeatures(obj_filtered), 10)\n",
    "  \n",
    "  # plot variable features with and without labels\n",
    "  plot1 <- VariableFeaturePlot(obj_filtered)\n",
    "  plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)\n",
    "  p1 <- plot1 + plot2\n",
    "  ggsave(plot = p1, filename = paste0(\"s4.\", sampleName, \"_\", supercluster, \".VariableFeaturePlot.png\"),\n",
    "                                      width = 12, height = 6, dpi = 300)\n",
    "  \n",
    "  #scaling\n",
    "  #all.genes <- rownames(obj_filtered)\n",
    "  obj_filtered <- ScaleData(obj_filtered, vars.to.regress = c(\"nCount_RNA\", \"percent.mt\"))\n",
    "  \n",
    "  #PCA\n",
    "  obj_filtered <- RunPCA(obj_filtered, features = VariableFeatures(object = obj_filtered))\n",
    "  p2 <- ElbowPlot(obj_filtered, ndims = 50)\n",
    "  ggsave(plot = p2, filename = paste0(\"s4.\", sampleName, \"_\", supercluster, \".ElbowPlot.png\"),\n",
    "         width = 10, height = 6, dpi = 300)\n",
    "  \n",
    "  #Clustering\n",
    "  obj_filtered <- FindNeighbors(obj_filtered, dims = 1:nPCs)\n",
    "  obj_filtered <- FindClusters(obj_filtered, resolution = resolution)\n",
    "  \n",
    "  #UMAP\n",
    "  obj_filtered <- RunUMAP(obj_filtered, dims = 1:nPCs)\n",
    "  obj_filtered <- RunTSNE(obj_filtered, dims = 1:nPCs)\n",
    "  p3 <- DimPlot(obj_filtered, reduction = \"umap\")\n",
    "  ggsave(plot = p3, filename = paste0(\"s4.\", sampleName, \"_\", supercluster, \".DimPlot.png\"),\n",
    "         width = 6, height = 6, dpi = 300)\n",
    "  \n",
    "  #export RDS\n",
    "  saveRDS(object = obj_filtered, file = paste0(sampleName, \"_\", supercluster, \"_PCs_\", nPCs, \"_\", resolution, \".RDS\")) \n",
    "}\n",
    "\n",
    "analyzeAndSaveObjects(seu_obj = wt12, supercluster = \"Myeloid\", \n",
    "                      nPCs = 14, resolution = 0.5, sampleName = \"wt12\")\n",
    "analyzeAndSaveObjects(seu_obj = dare12, supercluster = \"Myeloid\", \n",
    "                      nPCs = 21, resolution = 0.5, sampleName = \"dare12\")\n",
    "\n",
    "analyzeAndSaveObjects(seu_obj = wt12, supercluster = \"Lymphoid\", \n",
    "                      nPCs = 15, resolution = 0.5, sampleName = \"wt12\")\n",
    "analyzeAndSaveObjects(seu_obj = dare12, supercluster = \"Lymphoid\", \n",
    "                      nPCs = 23, resolution = 0.5, sampleName = \"dare12\")\n",
    "\n",
    "analyzeAndSaveObjects(seu_obj = wt12, supercluster = \"Stroma\", \n",
    "                      nPCs = 20, resolution = 0.5, sampleName = \"wt12\")\n",
    "analyzeAndSaveObjects(seu_obj = dare12, supercluster = \"Stroma\", \n",
    "                      nPCs = 22, resolution = 0.5, sampleName = \"dare12\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step5_DoubletFinder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```r\n",
    "setwd(\"D:/01_study/03_scRNA-seq/Iliopoulou_et_Al_Wt_Dare12wk_sc_analysis/analysis/Step5_DoubletFinder/\")\n",
    "rm(list=ls())\n",
    "library(Seurat)\n",
    "library(DoubletFinder)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "\n",
    "## Pre-process Seurat object (standard) --------------------------------------------------------------------------------------\n",
    "seu <- readRDS(\"../Step4_Analysis_of_Stroma_Lymphoid_Myeloid/wt12_Stroma_PCs_20_0.5.RDS\")\n",
    "\n",
    "## pK Identification (no ground-truth) ---------------------------------------------------------------------------------------\n",
    "sweep.res.list_seu <- paramSweep(seu, PCs = 1:20, sct = FALSE)\n",
    "sweep.stats_seu <- summarizeSweep(sweep.res.list_seu, GT = FALSE)\n",
    "bcmvn_seu <- find.pK(sweep.stats_seu)\n",
    "\n",
    "bcmvn_seu %>%\n",
    "  ggplot( aes(x=pK, y=BCmetric, group=1)) +\n",
    "  geom_line( color=\"grey\") +\n",
    "  geom_point(shape=21, color=\"black\", fill=\"#69b3a2\", size=2) +\n",
    "  theme_bw() +\n",
    "  ggtitle(\"pK selection\")\n",
    "\n",
    "## Homotypic Doublet Proportion Estimate -------------------------------------------------------------------------------------\n",
    "homotypic.prop <- modelHomotypic(seu@meta.data$RNA_snn_res.0.5)\n",
    "nExp_poi <- round(0.046*nrow(seu@meta.data))\n",
    "nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))\n",
    "\n",
    "## Run DoubletFinder with varying classification stringencies ----------------------------------------------------------------\n",
    "seu <- doubletFinder(seu, PCs = 1:20, pN = 0.25, pK = 0.3, nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)\n",
    "seu <- doubletFinder(seu, PCs = 1:20, pN = 0.25, pK = 0.3, nExp = nExp_poi.adj, reuse.pANN = \"pANN_0.25_0.3_453\", sct = FALSE)\n",
    "\n",
    "seu$DF.classifications_0.25_0.3_453 <- gsub(x = seu$DF.classifications_0.25_0.3_453, pattern = \"Doublet\", replacement = TRUE)\n",
    "seu$DF.classifications_0.25_0.3_453 <- gsub(x = seu$DF.classifications_0.25_0.3_453, pattern = \"Singlet\", replacement = FALSE)\n",
    "seu$DF.classifications_0.25_0.3_397 <- gsub(x = seu$DF.classifications_0.25_0.3_397, pattern = \"Doublet\", replacement = TRUE)\n",
    "seu$DF.classifications_0.25_0.3_397 <- gsub(x = seu$DF.classifications_0.25_0.3_397, pattern = \"Singlet\", replacement = FALSE)\n",
    "\n",
    "seu$DoubletFinderClass <- as.logical(seu$DF.classifications_0.25_0.3_453) | as.logical(seu$DF.classifications_0.25_0.3_397)\n",
    "\n",
    "DimPlot(seu, group.by = \"DoubletFinderClass\", split.by = \"DoubletFinderClass\")\n",
    "saveRDS(seu, \"wt12_Stroma_PCs_20_0.05_DoubletFinder.RDS\")\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "rm(list=ls(all=TRUE))#remove previous vars, objects\n",
    "\n",
    "## Pre-process Seurat object (standard) --------------------------------------------------------------------------------------\n",
    "seu <- readRDS(\"../Step4_Analysis_of_Stroma_Lymphoid_Myeloid/dare12_Stroma_PCs_22_0.5.RDS\")\n",
    "\n",
    "## pK Identification (no ground-truth) ---------------------------------------------------------------------------------------\n",
    "sweep.res.list_seu <- paramSweep(seu, PCs = 1:22, sct = FALSE)\n",
    "sweep.stats_seu <- summarizeSweep(sweep.res.list_seu, GT = FALSE)\n",
    "bcmvn_seu <- find.pK(sweep.stats_seu)\n",
    "\n",
    "bcmvn_seu %>%\n",
    "  ggplot( aes(x=pK, y=BCmetric, group=1)) +\n",
    "  geom_line( color=\"grey\") +\n",
    "  geom_point(shape=21, color=\"black\", fill=\"#69b3a2\", size=2) +\n",
    "  theme_bw() +\n",
    "  ggtitle(\"pK selection\")\n",
    "\n",
    "## Homotypic Doublet Proportion Estimate -------------------------------------------------------------------------------------\n",
    "homotypic.prop <- modelHomotypic(seu@meta.data$RNA_snn_res.0.5)           \n",
    "nExp_poi <- round(0.046*nrow(seu@meta.data))\n",
    "nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))\n",
    "\n",
    "## Run DoubletFinder with varying classification stringencies ----------------------------------------------------------------\n",
    "seu <- doubletFinder(seu, PCs = 1:22, pN = 0.25, pK = 0.25, nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)\n",
    "seu <- doubletFinder(seu, PCs = 1:22, pN = 0.25, pK = 0.25, nExp = nExp_poi.adj, reuse.pANN = \"pANN_0.25_0.25_220\", sct = FALSE)\n",
    "\n",
    "seu$DF.classifications_0.25_0.25_220 <- gsub(x = seu$DF.classifications_0.25_0.25_220, pattern = \"Doublet\", replacement = TRUE)\n",
    "seu$DF.classifications_0.25_0.25_220 <- gsub(x = seu$DF.classifications_0.25_0.25_220, pattern = \"Singlet\", replacement = FALSE)\n",
    "seu$DF.classifications_0.25_0.25_191 <- gsub(x = seu$DF.classifications_0.25_0.25_191, pattern = \"Doublet\", replacement = TRUE)\n",
    "seu$DF.classifications_0.25_0.25_191 <- gsub(x = seu$DF.classifications_0.25_0.25_191, pattern = \"Singlet\", replacement = FALSE)\n",
    "\n",
    "seu$DoubletFinderClass <- as.logical(seu$DF.classifications_0.25_0.25_220) | as.logical(seu$DF.classifications_0.25_0.25_191)\n",
    "\n",
    "DimPlot(seu, group.by = \"DoubletFinderClass\", split.by = \"DoubletFinderClass\")\n",
    "saveRDS(seu, \"dare12_Stroma_PCs_22_0.05_DoubletFinder.RDS\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step6_Integration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```R\n",
    "# 这份代码展示了单细胞RNA测序数据分析中一个非常关键的步骤——多样本数据整合（Integration）​。\n",
    "# 简单来说，它的核心目标是消除因实验批次、不同样本来源（如野生型WT vs 基因编辑型DARE）\n",
    "# 带来的技术差异，使得不同样本间相同类型的细胞能够聚集在一起，\n",
    "# 从而让我们能够真实地比较生物学条件（如基因型）对细胞产生的影响\n",
    "\n",
    "setwd(\"D:/01_study/03_scRNA-seq/Iliopoulou_et_Al_Wt_Dare12wk_sc_analysis/analysis/Step6_Integration/\")\n",
    "rm(list=ls())\n",
    "library(Seurat)\n",
    "\n",
    "#Load objects\n",
    "# 代码首先加载了6个已经过前期处理的Seurat对象，\n",
    "# 它们分别来自野生型（WT）​​ 和 ​基因编辑型（DARE）​​ 小鼠样本，\n",
    "# 并已根据主要细胞谱系（淋巴系Lym、髓系Mye、基质Str）进行了分群。\n",
    "# 这种分细胞大类进行独立整合的策略非常关键，它能有效避免整合算法将不同谱系的细胞错误地“拉近”，\n",
    "# 从而更好地保留细胞亚群间的生物学差异\n",
    "dareLym <- readRDS(\"../Step5_DoubletFinder/dare12_Lymphoid_PCs_23_0.05_DoubletFinder.RDS\")\n",
    "dareMye <- readRDS(\"../Step5_DoubletFinder/dare12_Myeloid_PCs_21_0.5_DoubletFinder.RDS\")\n",
    "dareStr <- readRDS(\"../Step5_DoubletFinder/dare12_Stroma_PCs_22_0.05_DoubletFinder.RDS\")\n",
    "\n",
    "wtLym <- readRDS(\"../Step5_DoubletFinder/wt12_Lymphoid_PCs_15_0.05_DoubletFinder.RDS\")\n",
    "wtMye <- readRDS(\"../Step5_DoubletFinder/wt12_Myeloid_PCs_14_0.5_DoubletFinder.RDS\")\n",
    "wtStr <- readRDS(\"../Step5_DoubletFinder/wt12_Stroma_PCs_20_0.05_DoubletFinder.RDS\")\n",
    "\n",
    "integrateSeuratObjects <- function(obj1, obj2, nPCs, resolution, exportName)\n",
    "{\n",
    "  #change Idents and keep only unique cells\n",
    "  Idents(obj1) <- obj1$DoubletFinderClass\n",
    "  Idents(obj2) <- obj2$DoubletFinderClass\n",
    "  obj1 <- subset(obj1, idents = \"FALSE\")\n",
    "  obj2 <- subset(obj2, idents = \"FALSE\")\n",
    "  \n",
    "  #pre-processing \n",
    "  obj.list <- list(obj1, obj2)\n",
    "  # normalize and identify variable features for each dataset independently\n",
    "  # 1. 将两个对象放入列表，并分别对每个进行标准化和寻找高变基因\n",
    "  obj.list <- lapply(X = obj.list, FUN = function(x) {\n",
    "    x <- NormalizeData(x)\n",
    "    x <- FindVariableFeatures(x, selection.method = \"vst\", nfeatures = 2000)\n",
    "  })\n",
    "  \n",
    "  # select features that are repeatedly variable across datasets for integration\n",
    "  features <- SelectIntegrationFeatures(object.list = obj.list)\n",
    "  \n",
    "  #Perform integration\n",
    "  obj.anchors <- FindIntegrationAnchors(object.list = obj.list, anchor.features = features)\n",
    "  # FindIntegrationAnchors函数会寻找两个数据集中那些处于相同生物学状态的细胞对（即“锚点”）。\n",
    "  # 这些锚点就像是不同地图上的共同地标，后续的整合算法会以它们为参考，\n",
    "  # 将不同数据集（样本）的细胞正确“对齐”到同一个坐标系中，从而校正批次效应。\n",
    "  \n",
    "  # this command creates an 'integrated' data assay\n",
    "  obj.combined <- IntegrateData(anchorset = obj.anchors)\n",
    "  \n",
    "  #Perform integrated analysis\n",
    "  # specify that we will perform downstream analysis on the corrected data note that the\n",
    "  # original unmodified data still resides in the 'RNA' assay\n",
    "  DefaultAssay(obj.combined) <- \"integrated\"\n",
    "  # 整合后的数据存储在名为 \"integrated\"的 assay 中，而原始数据仍保留在 \"RNA\"assay 里。\n",
    "  # 后续的所有降维和聚类操作都基于校正后的数据，\n",
    "  # 这能确保聚类结果由真实的生物学状态驱动，而非技术偏差\n",
    "  \n",
    "  # Run the standard workflow for visualization and clustering\n",
    "  obj.combined <- ScaleData(obj.combined, verbose = T)\n",
    "  obj.combined <- RunPCA(obj.combined, npcs = 30, verbose = T)\n",
    "  ElbowPlot(obj.combined, ndims = 30)\n",
    "  obj.combined <- RunUMAP(obj.combined, reduction = \"pca\", dims = 1:nPCs)\n",
    "  obj.combined <- FindNeighbors(obj.combined, reduction = \"pca\", dims = 1:nPCs)\n",
    "  obj.combined <- FindClusters(obj.combined, resolution = 0.5)\n",
    "  \n",
    "  #save RDS\n",
    "  saveRDS(object = obj.combined, file = exportName)\n",
    "}\n",
    "\n",
    "# 代码最后分别对基质、淋巴和髓系细胞进行了三次独立的整合操作。\n",
    "# 值得注意的是，每次整合都使用了不同的主成分数（nPCs）​​（23, 15, 20），\n",
    "# 这可能是研究者根据每个细胞大类数据的复杂程度（通过ElbowPlot判断）进行的个性化设置，\n",
    "# 目的是捕捉最主要的生物学变异\n",
    "\n",
    "integrateSeuratObjects(obj1 = wtStr, obj2 = dareStr, nPCs = 23, resolution = 0.5, exportName = \"WT12wk_DARE12Wk_Integration_Stroma_VST_without_DoubletsFromDoubletFinder.RDS\")\n",
    "integrateSeuratObjects(obj1 = wtLym, obj2 = dareLym, nPCs = 15, resolution = 0.5, exportName = \"WT12wk_DARE12Wk_Integration_Lymphoid_VST_without_DoubletsFromDoubletFinder.RDS\")\n",
    "integrateSeuratObjects(obj1 = wtMye, obj2 = dareMye, nPCs = 20, resolution = 0.5, exportName = \"WT12wk_DARE12Wk_Integration_Myeloid_VST_without_DoubletsFromDoubletFinder.RDS\")\n",
    "\n",
    "# 这段代码实现了一个稳健、高效的单细胞数据整合流程。其最大价值在于：\n",
    "# ​消除批次效应​：通过CCA等方法识别生物学上的“锚点”，\n",
    "#     有效校正了不同样本（WT vs DARE）之间的技术差异。\n",
    "# ​实现可比性​：使得我们能够在一个统一的空间中，\n",
    "#     公平地比较不同基因型背景下，同一细胞类型的基因表达差异，为发现真正的生物学现象奠定了基础\n",
    "# 希望这份详细的解读能帮助你深入理解单细胞数据整合的原理和操作！\n",
    "# 如果你对某个具体步骤还有疑问，我们可以继续探讨"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step7_Annotation_of_clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```r\n",
    "\n",
    "setwd(\"D:/01_study/03_scRNA-seq/Iliopoulou_et_Al_Wt_Dare12wk_sc_analysis/analysis/Step7_Annotation_of_clusters/\")\n",
    "rm(list=ls())\n",
    "\n",
    "library(Seurat)\n",
    "library(scCustomize)\n",
    "library(ggplot2)\n",
    "\n",
    "#-----------------------Annotation of STROMA cells------------------------------\n",
    "stroma <- readRDS(\"../Step6_Integration/WT12wk_DARE12Wk_Integration_Stroma_VST_without_DoubletsFromDoubletFinder.RDS\")\n",
    "DefaultAssay(stroma) <- \"integrated\"\n",
    "\n",
    "#Exclusion of clusters 10, 14, 19\n",
    "stroma <- subset(x = stroma, idents = c(10, 14, 19), invert = T)\n",
    "stroma <- FindNeighbors(stroma, reduction = \"pca\", dims = 1:23)\n",
    "stroma <- FindClusters(stroma, resolution = 0.4)\n",
    "\n",
    "DimPlot(stroma, reduction = \"umap\")\n",
    "FeaturePlot(stroma, features = \"Pdgfra_low\", order = T, \n",
    "            label = T, cols = c(\"grey70\", \"red\"))\n",
    "\n",
    "#Annotation of clusters after manual inspection of marker genes from the literature\n",
    "new.cluster.ids <- c(\"Pdgfra_low\",\"Pdgfra_low\", \"LECS\", \"Trophocytes\", \"Telocytes\", \n",
    "                     \"SMCs\",\"FRC2\", \"FRC1\", \"Capillary_ECs\", \"Trophocytes\", \"Pericytes\", \n",
    "                     \"FDCs_MRCs\",\"Cajal\", \"Vein_ECs\", \"Mesothelial\", \"Glial\")\n",
    "names(new.cluster.ids) <- levels(stroma)\n",
    "stroma <- RenameIdents(object = stroma, new.cluster.ids)\n",
    "stroma$annotation <- Idents(stroma)\n",
    "saveRDS(stroma, \"WT12wk_DARE12Wk_Stroma_annotated.RDS\")\n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "#-----------------------Annotation of Lymphoid cells----------------------------\n",
    "lymphoid <- readRDS(\"../Step6_Integration/WT12wk_DARE12Wk_Integration_Lymphoid_VST_without_DoubletsFromDoubletFinder.RDS\")\n",
    "DefaultAssay(lymphoid) <- \"integrated\"\n",
    "lymphoid <- FindClusters(lymphoid, resolution = 0.7)\n",
    "DimPlot_scCustom(lymphoid)\n",
    "\n",
    "#Exclusion of cluster 16\n",
    "lymphoid <- subset(lymphoid, idents = 16, invert = T)\n",
    "DimPlot_scCustom(lymphoid)\n",
    "lymphoid\n",
    "lymphoid <- FindNeighbors(lymphoid, reduction = \"pca\", dims = 1:23)\n",
    "lymphoid <- FindClusters(lymphoid, resolution = 0.7)\n",
    "\n",
    "#Annotation of clusters after manual inspection of marker genes from the literature\n",
    "new.cluster.ids <- c(\"B_cells\",\"Plasma_cells\",\"Plasma_cells\",\"Plasma_cells\", \"Memory_T_cells\", \"ILC2s\", \"Naïve_CD4_T_cells\", \n",
    "                     \"Cd8_T_cells\", \"Th17_cells\", \"ILC3s\", \"Tregs\", \"Cycling_B_cells\", \"Tcr_gammadelta\", \"B_cells\",\"Naïve_CD8_T_cells\",\n",
    "                     \"Cycling_T_cells\", \"Plasma_cells\",\"NKT_cells\",\"Memory_T_cells\", \"B_cells\")\n",
    "names(new.cluster.ids) <- levels(lymphoid)\n",
    "lymphoid <- RenameIdents(lymphoid, new.cluster.ids)\n",
    "lymphoid$annotation <- Idents(lymphoid)\n",
    "saveRDS(lymphoid, \"WT12wk_DARE12Wk_Lymphoid_annotated.RDS\")\n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "#-----------------------Annotation of Myeloid cells-----------------------------\n",
    "myeloid <- readRDS(\"../Step6_Integration/WT12wk_DARE12Wk_Integration_Myeloid_VST_without_DoubletsFromDoubletFinder.RDS\")\n",
    "DefaultAssay(myeloid) <- \"integrated\"\n",
    "myeloid <- FindClusters(myeloid, resolution = 0.8)\n",
    "DimPlot_scCustom(myeloid)\n",
    "\n",
    "#Exclusion of clusters 11, 12, 13, 14, 16\n",
    "myeloid <- subset(myeloid, idents = c(\"11\", \"12\", \"13\", \"14\", \"16\"), invert = T)\n",
    "myeloid <- FindNeighbors(myeloid, reduction = \"pca\", dims = 1:23)\n",
    "myeloid <- FindClusters(myeloid, resolution = 0.5)\n",
    "\n",
    "#Annotation of clusters after manual inspection of marker genes from the literature\n",
    "new.cluster.ids <- c(\"Intermediate_Mfs\",\"Granulocytes\", \"Granulocytes\", \"Activated_Mfs\", \"Resident_Mfs\",\n",
    "                     \"Cdc2_a\",\"Monocytes\", \"Cdc1\", \"Mixed_Mf_popul\", \"Cdc2_b\", \"pDCs\", \"Cdc2_b\")\n",
    "names(new.cluster.ids) <- levels(myeloid)\n",
    "myeloid <- RenameIdents(myeloid, new.cluster.ids)\n",
    "\n",
    "#exclusion of a mixed MF population\n",
    "myeloid <- subset(myeloid, idents = c(\"Mixed_Mf_popul\"), invert = T)\n",
    "myeloid <- FindNeighbors(myeloid, reduction = \"pca\", dims = 1:23)\n",
    "DimPlot_scCustom(myeloid)\n",
    "myeloid$annotation <- Idents(myeloid)\n",
    "saveRDS(myeloid, \"WT12wk_DARE12Wk_Myeloid_annotated.RDS\")\n",
    "#-------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "该流程从读取两个数据，一个野生型和一个实验型，分别进行质控、降维、聚类、整合；但因为在细胞注释时，代码中没有该处详细的注释，如何因为软件的差异导致注释不一致，导致代码运行出现问题。所以后面没有进行跟进follow了。详细可以见开头的GitHub链接。"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "10x",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
